Apply colors only when they are supported

nikableh opened the pull request:
Closes #236.
Closes #237.

This PR wraps all calls to `owo_colors::OwoColorize::style` with function that calls `anstream::AutoStream::choice` and checks whenever colors are available for the current stderr.

### Conserns

The stderr here is hardcoded it that somewhat bothers me. It is possible for that user of eyre to configure printing all of the error messages into the stderr? Because if eyre ever prints color to the stdout and then the user redirects stdout to the file, [this check](https://github.com/eyre-rs/eyre/blob/99576ecac083080e9668472b2d7b18e301ec2ad0/color-spantrace/src/lib.rs#L392) would detect that stderr is still a terminal and allow ANSI symbols to be printed to the file. 

I am also wondering if duplicating `style_if_possible` in `color-spantrace` and `color-eyre` is okay. Should I move it somewhere else?

### Tests

I've done some quick testing by running examples and this seems to work well. I don't see any colors in the terminal and there are no special symbols _from eyre_ when redirecting to the file.

Here is a script to run all color-eyre examples and print the output to the file. Mind that `RUSTFLAGS=-Awarnings` will force everything to rebuild. But that saves you from scrolling past all the deprecation warnings (fixed here btw — https://github.com/eyre-rs/eyre/pull/235).

```bash
#!/bin/env bash

TEST_OUTPUT_DIR=test-no-color-examples
EXAMPLES='custom_filter custom_section debug_perf github_issue multiple_errors panic_compose panic_hook theme_test_helper theme color-eyre-usage'

rm -r $TEST_OUTPUT_DIR || true
mkdir -p $TEST_OUTPUT_DIR

for example in $EXAMPLES; do
    RUSTFLAGS=-Awarnings cargo run --example $example &>> $TEST_OUTPUT_DIR/$example
done
```

Might be nice to write some unit tests, but I'd prefer if someone else did it.

### About tracing_subscriber

The only ANSI symbols I see are the ones coming from tracing_subscriber ([color-eyre/examples/theme_test_helper.rs:35](https://github.com/eyre-rs/eyre/blob/master/color-eyre/examples/theme_test_helper.rs#L35)). I did a little digging and found that removing this layer [color-eyre/examples/theme_test_helper.rs:49](https://github.com/eyre-rs/eyre/blob/master/color-eyre/examples/theme_test_helper.rs#L49) from registry fixes the issue.

The colors are still there, in the terminal, but not in the file. The only thing that changed visually is that the text is no longer cursive. Although I don't quite understand what this layer does and why is it needed. The [documentation](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/struct.Layer.html#method.with_target) is not clear to me.

I think we should just remove this layer from every example to avoid teaching people this flawed behaviour. Let's leave it up to them to enable this layer. This way it won't be eyre's problem, but tracing_subscriber's one.

### Rant about owo_colors

To be fair, this all feels like a huge blunder from owo_colors. I feel like it is it who should be responsible for detecting colors. They already have a `supports-color` feature. But enabling it only gives you a method that you should apply everywhere yourself. I can understand why this feature isn't turned on by default, sure. But once I enable it I kind of expect everything to work out of the box. What if you have thousands of lines of code where colors are added? Do you wrap everything with `if_supports_color`??? This is nonsense.

nikableh committed:
Apply colors only when they are supported (#236, #237)

pksunkara reviewed (changes requested):
Ideally, we can just call the `style_if_possible` function as `style`. From a developer's perspective, they don't care whether that function is styling or not, depending on the env vars.

pksunkara reviewed (changes requested):
Can you please do the following changes:

1. Replace all `owo_colors::style()` usage with `owo_colors::Style::new()` (needed for 2nd change)
2. Rename all `style_if_possible` to `style`.

epage commented:
`anstream` is a larger dependency for just determining color choice and has specific logic dealing with its auto-adapting to old versions of Windows.

You could pull out the choice logic, calling `anstyle_query` directly, and adapt to the needs here.

Or eyre could do something like clap and be neutral to this and assume the caller is using `anstream` with ways of stripping the color.

nikableh committed:
Replace `owo_colors::style()` with `owo_colors::Style::new()`

nikableh committed:
Replace `owo_colors::style()` with `owo_colors::Style::new()`

nikableh committed:
Rename `crate::style_if_possible` into `crate::style`

pksunkara reviewed (approved):
Looks good. I will merge once you can get CI passing.

aditeyapatakoti reviewed (approved):
nice edits

pksunkara commented:
@synalice Can you please fix CI?

pksunkara commented:
Still failing. Can you please try all the commands from CI locally first? Using github actions takes a lot of time since a maintainer needs to approve every time.

pksunkara reviewed (changes requested):
Looking at the failing tests, you need to enable colors in tests so that they can be tested.

nikableh committed:
Apply colors only when they are supported (#236, #237)

nikableh committed:
Force colors on in backwards-compatibility tests

Since cce2d77, colors are emitted only when the output stream supports
them. The backwards-compatibility tests compare freshly formatted
output against colored control data (`theme_control.txt`). Under `cargo
test` their stderr is a pipe, not a tty, so color is auto-disabled and
the generated output no longer matches the control data.

Set `CLICOLOR_FORCE=1` so colors are emitted regardless of stream
support, keeping the comparisons valid.

nikableh committed:
Fix clippy errors

pksunkara commented on Cargo.toml:
```diff
@@ -20,6 +20,7 @@ include = ["README.md", "LICENSE-APACHE", "LICENSE-MIT", "build.rs", "Cargo.toml
 indenter = "0.3.0"
 owo-colors = "4.3"
 autocfg = "1.0"
+anstream = "0.6.15"
```
[SUGGESTION] Let's update this to v1

pksunkara commented on color-eyre/examples/theme.rs:
```diff
@@ -1,12 +1,13 @@
-use color_eyre::{Section, config::Theme, eyre::Report, owo_colors::style};
+use color_eyre::{Section, config::Theme, eyre::Report};
+use owo_colors::Style;
```
[SUGGESTION] Since we re-export `owo-colors`, we should probably recommend them to import it from `color_eyre`. So, keep the previous.

pksunkara commented on color-eyre/src/lib.rs:
```diff
@@ -361,13 +361,16 @@ use std::sync::Arc;
 #[doc(hidden)]
 pub use eyre::Result;
 pub use owo_colors;
+use owo_colors::OwoColorize;
+use owo_colors::Style;
```
[NOTE] Combine these 2 lines into a single one.

pksunkara commented on color-eyre/src/lib.rs:
```diff
@@ -458,3 +461,17 @@ pub enum ErrorKind<'a> {
+where
+    S: ToString + std::fmt::Display,
+{
+    let color_choice = AutoStream::choice(&std::io::stderr());
```
[ISSUE] Instead of deriving this every call, we should probably do this once at the beginning (maybe during `install`) and use it when called.

pksunkara commented on color-spantrace/src/lib.rs:
```diff
@@ -374,3 +375,17 @@ impl fmt::Display for ColorSpanTrace<'_> {
+where
+    S: ToString + std::fmt::Display,
+{
+    let color_choice = AutoStream::choice(&std::io::stderr());
```
[ISSUE] 1. If we are deriving this here, why are we doing this again in `color_eyre`? Is it not better to unify both?
2. Is there no way to cache this?

This is a major blocker to finalize this PR. Since I will be pretty active, let's resolve this and merge it.

pksunkara reviewed (changes requested):
[ISSUE] Please add a small new test to color-eyre where we test with colors absent.

<sub>Reviewed with [tuicr](https://github.com/agavra/tuicr).</sub>
